File: src/ui/library.js
(function(){
  window.UI = window.UI || {};
  window.UI.library = function(container, state, {addXP, addItem, completeQuest}){
    function render(){
      container.innerHTML = '';
      const root = document.createElement('div');
      root.className = 'ui-grid';

      const card = document.createElement('div');
      card.className = 'ui-card';
      const title = document.createElement('div'); title.className = 'ui-title'; title.textContent = 'Library Services';
      card.appendChild(title);

      if (!state.flags.libraryCard) {
        const row = document.createElement('div'); row.className = 'hstack';
        const btn = document.createElement('button'); btn.className = 'btn primary'; btn.textContent = 'Register for Library Card (+20 XP)';
        btn.addEventListener('click', function(){ if(state.flags.libraryCard) return; state.flags.libraryCard = true; addItem('Library Card'); addXP(20); completeQuest('libraryVisit'); render(); });
        row.appendChild(btn);
        card.appendChild(row);
        root.appendChild(card);
        container.appendChild(root);
        return;
      }

      const booksCard = document.createElement('div');
      booksCard.className = 'ui-card';
      const t2 = document.createElement('div'); t2.className = 'ui-title'; t2.textContent = 'Browse Books';
      booksCard.appendChild(t2);

      const categories = {
        Engineering: ['Statics Basics', 'Digital Logic', 'Fluid Flow 101'],
        Science: ['Physics Primer', 'Organic Chemistry', 'Biology of Cells'],
        Literature: ['Poetry Classics', 'Modern Novels', 'World Myths']
      };

      const catGrid = document.createElement('div'); catGrid.className = 'ui-grid';
      Object.entries(categories).forEach(([cat, list]) => {
        const c = document.createElement('div'); c.className = 'ui-card';
        const h = document.createElement('div'); h.className = 'ui-title'; h.textContent = cat; c.appendChild(h);
        list.forEach(title => {
          const row = document.createElement('div'); row.className = 'hstack';
          const lbl = document.createElement('div'); lbl.textContent = title; row.appendChild(lbl);
          const btn = document.createElement('button'); btn.className = 'btn'; btn.textContent = 'Read (+2 XP)';
          btn.addEventListener('click', function(){ state.meta.booksRead = (state.meta.booksRead||0)+1; addXP(2); });
          row.appendChild(btn);
          c.appendChild(row);
        });
        catGrid.appendChild(c);
      });
      booksCard.appendChild(catGrid);

      const stats = document.createElement('div'); stats.className = 'ui-card';
      const t3 = document.createElement('div'); t3.className = 'ui-title'; t3.textContent = 'Reading Stats';
      const s = document.createElement('div'); s.className = 'ui-subtle'; s.textContent = 'Books read: ' + (state.meta.booksRead||0);
      stats.appendChild(t3); stats.appendChild(s);

      root.appendChild(card);
      root.appendChild(booksCard);
      root.appendChild(stats);
      container.appendChild(root);
    }

    render();
  };
})();

File: python/game/ui/library.py
from ..popup import Popup

CATEGORIES = {
    'Engineering': ['Statics Basics', 'Digital Logic', 'Fluid Flow 101'],
    'Science': ['Physics Primer', 'Organic Chemistry', 'Biology of Cells'],
    'Literature': ['Poetry Classics', 'Modern Novels', 'World Myths']
}


def build_popup(h):
    s = h['state']
    p = Popup("Library")

    if not s.flags['libraryCard']:
        p.add_line("Register for a library card to access books.")
        def register():
            if s.energy < 10:
                h['toast']("Too tired! Need at least 10 energy.")
                return
            s.flags['libraryCard'] = True
            h['addItem']('Library Card')
            h['addXP'](20)
            h['addReputation'](5)  # Getting library card is a responsible action
            h['addEnergy'](-10)
            h['completeQuest']('libraryVisit')
            h['toast']("Library card registered (+20 XP, +5 Reputation)")
            h['rebuild']()
        p.add_button("Register (+20 XP)", register, primary=True)
        return p

    p.add_line("Browse and read books:")
    # Only one action at a time in this compact popup; rotate categories
    def make_read(title):
        def _read():
            if s.energy < 15:
                h['toast']("Too tired to read! Need at least 15 energy.")
                return
            s.meta['booksRead'] = int(s.meta.get('booksRead', 0)) + 1
            h['addXP'](2)
            knowledge_gain = 6
            # High knowledge makes reading more effective
            if s.knowledge > 50:
                knowledge_gain += 2
            h['addKnowledge'](knowledge_gain)
            h['addStress'](-2)  # Reading is relaxing
            h['addEnergy'](-15)
            h['addDiscipline'](2)
            h['toast'](f'Read: {title} (+2 XP, +{knowledge_gain} Knowledge, -2 Stress)')
            h['rebuild']()
        return _read

    added = 0
    for cat, books in CATEGORIES.items():
        for title in books:
            if added >= 3:
                break
            p.add_button(f"Read: {title}", make_read(title), primary=False)
            added += 1
        if added >= 3:
            break

    return p
